crates/burn-jit/src/kernel/matmul/tiling2d_shader/load_shared_memory.rs

```diff
@@ -0,0 +1,271 @@
+        ),
+    };
+    let k = shader_state.k;
+    let dim_k = shader_state.dim_k;
+
+    // How close is the thread to the end of the matrix.
+    // If < 4 then it is an edge case
+    let remain = scope.create_local(Elem::UInt);
+    gpu!(scope, remain = dim - pos_in_dim);
+
+    let block_size_k: Variable = shader.config.block_size_k.into();
+    let block_size_n: Variable = shader.config.block_size_n.into();
+    let elem = input.item().elem();
+
+    let current = scope.create_local(Elem::UInt);
+    let aligned_with_shared_memory = scope.create_local(Elem::Bool);
+    let sm_position = scope.create_local(Elem::UInt);
+    let within_input = scope.create_local(Elem::Bool);
+    let current_with_k = scope.create_local(Elem::UInt);
+    let remain_at_least_1 = scope.create_local(Elem::Bool);
+    let read_condition = scope.create_local(Elem::Bool);
+    let val_vec4 = scope.create_local(shared_memory.item());
+
+    gpu!(
+        scope,
+        range(0_u32, 4u32, shader.unroll).for_each(|j, scope| {
+            gpu!(scope, current = thread_idx_1 + j);
+
+            gpu!(scope, aligned_with_shared_memory = current < block_size_k);
+
+            // To avoid overwriting following row in shared memory
+            gpu!(scope, if(aligned_with_shared_memory).then(|scope|{
+
+                // Position in shared memory
+                match input_identifier {
+                    InputIdentifier::Lhs => {
+                        gpu!(scope, sm_position = thread_idx_2 / 4u32);
+                        gpu!(scope, sm_position *= block_size_k);
+                        gpu!(scope, sm_position += current);
+                },
+                    InputIdentifier::Rhs => {
+                        gpu!(scope, sm_position = current * block_size_n);
+                        gpu!(scope, sm_position += thread_idx_2);
+                        gpu!(scope, sm_position = sm_position / 4u32);
+                    }
+                }
+
+                // To pad with zeros if outside lhs
+                gpu!(scope, current_with_k = current + k);
+                gpu!(scope, within_input = current_with_k < dim_k);
+                gpu!(scope, remain_at_least_1 = remain >= 1u32);
+                gpu!(scope, read_condition = within_input && remain_at_least_1);
+
+                gpu!(scope, if(read_condition).then(|scope| {
+                    let tmp = scope.create_local(Elem::UInt);
+                    let position_0 = scope.create_local(Elem::UInt);
+                    let position_1 = scope.create_local(Elem::UInt);
+                    let position_2 = scope.create_local(Elem::UInt);
+                    let position_3 = scope.create_local(Elem::UInt);
+                    let remain_n = scope.create_local(Elem::Bool);
+
+                    let val_0 = scope.zero(elem);
+                    let val_1 = scope.zero(elem);
+                    let val_2 = scope.zero(elem);
+                    let val_3 = scope.zero(elem);
```

Comment:
Can we extract those else where, the declaration is probably done in a for loop. Also, not sure we need to initialize val_0, val_1, val_2 and val_3 to zero.